joomla/joomla-cms
# [4.0] Fix database for Extensions - GSoC Expand Extensions Manager

### Introduction

Currently when you have a problem in the Core Database (com_admin / installation tables) if there is any error you can fix it in the Extensions -> Manage -> Database, where it can show if you have any error in the database and which is.
The same doesn’t happen when you install a extension and you have problem on it, with this feature you can fix the extension without needing to re-install or even use queries in the joomla database.


### Summary of Changes

Improvement of the Extensions -> Manage -> Database view for fixing now not only the core Joomla database but also 3rd party extensions

### Testing Instructions

You can test with any feature:

1. Try to change the version_id of an extension in #__schemas 
2. Try to change the version in manifest_cache in #__extensions
3. Try to drop any table related to your extension or change any column

To check the problems get your pointer over in the Errors column, a tooltip will pop

Here is an example of my extension:
https://github.com/NunoLopes96/component_joomla

Use this query:

update `#__schemas` set version_id='0.0.12' where extension_id='10000';  (in my case)
drop table `#__students`;

After selecting my component the table came back again and the fix was successful

### Expected result

![screenshot from 2017-08-02 17-17-49](https://user-images.githubusercontent.com/20955884/28883484-8bc31ec2-77a6-11e7-859f-ed1e1e38abd0.png)

There will be 3 types of badge in Errors column in the table:
Success : There are no Problems on the database
Warning : There are only errors related to versions (Database schema or Update version) and non in the database.
Danger : There are errors on the database

For more information on the problems you can pass your mouse over the Errors:
![screenshot from 2017-08-02 14-53-21](https://user-images.githubusercontent.com/20955884/28883502-a2446b24-77a6-11e7-99bb-d8470086c707.png)

note: This extension has their Update Version and Database Schema version correct but still there are errors in the database

![screenshot from 2017-08-02 14-52-51](https://user-images.githubusercontent.com/20955884/28883513-a7273b80-77a6-11e7-8370-ed8342c1ce05.png)
note: There are no errors in the database, only the Database Schema version and the Update Version are incorrect, still you can fix this.


### Actual result

![screenshot from 2017-07-03 02-27-10](https://user-images.githubusercontent.com/20955884/27775154-23ec80d0-5f97-11e7-9f40-fed31f61a0f8.png)

### Documentation Changes Required
